### PR TITLE
Make ImageReference strongly-typed and checked up-front

### LINKED ISSUE #1464: Parse Azure Resource IDs into models
In `onefuzz.azure.image`, for example, we use `msrestazure.tools.parse_resource_id()` to get a `dict` representation of the components of an Azure Resource ID. This is _very_ dynamic and weakly-typed, and prone to absent keys in some situations, as a result of the `parse_resource_id()` function's return type and API.

Instead of doing ad hoc, implicit recognition of what kind of resource we are parsing, we should have models for the resource types we know about, grouped in a `Union`. 

See comment: https://github.com/microsoft/onefuzz/pull/1450/files#r751810404 

[AB#35946](https://fuzzfest.visualstudio.com/5eaacf23-2084-4d62-8fbe-7d37326a8b74/_workitems/edit/35946)

### LINKED ISSUE #2927: Azure Compute Gallery support no longer exists in the .NET release
## Information

- Onefuzz version: 6.4.0
- OS: Linux

## Provide detailed reproduction steps (if any)

The new OneFuzz .NET API does not properly handle Azure Compute Galleries. This is similar to an issue previously reported in the Python version of the API: https://github.com/microsoft/onefuzz/issues/1443

1. Specify a path to a compute gallery image in the instance config `default_linux_vm_image` setting. The issue will also occur if the image is specified using "--image" parameter when executing the `onefuzz create scalesets` command.

   The path provided is the Resource ID from the image definition's properties as defined by Azure. The format of the path is:

`/subscriptions/{SUBSCRIPTION_ID}/resourceGroups/{RESOURCE_GROUP}/providers/Microsoft.Compute/galleries/{GALLERY_NAME}/images/{IMAGE_NAME}/versions/{IMAGE_VERSION}`

The IMAGE_NAME is the name of the image (e.g. Custom_Ubuntu)
The IMAGE_VERSION is the specific version (e.g. 1.2.3)

2. Executing a `onefuzz scalesets create` command with a compute gallery image will result in a `creation_failed` state.  The following error message will be returned:

```
    {
        "ephemeral_os_disks": false,
        "error": {
            "code": "INVALID_IMAGE",
            "errors": [
 "Azure.RequestFailedException: The Resource 'Microsoft.Compute/images/{IMAGE_VERSION}' under resource group 'RESOURCE_GROUP' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix\r\nStatus: 404 (Not Found)\r\nErrorCode: ResourceNotFound\r\n\r\nContent:\r\n{\"error\":{\"code\":\"ResourceNotFound\",\"message\":\"The Resource 'Microsoft.Compute/images/{IMAGE_VERSION}' under resource group 'RESOURCE_GROUP' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix\"}}\r\n\r\nHeaders:\r\nCache-Control: no-cache\r\nPragma: no-cache\r\nx-ms-failure-cause: REDACTED\r\nx-ms-request-id: cd624acf-6361-446d-9829-5766e7c67210\r\nx-ms-correlation-request-id: REDACTED\r\nx-ms-routing-request-id: REDACTED\r\nStrict-Transport-Security: REDACTED\r\nX-Content-Type-Options: REDACTED\r\nDate: Thu, 16 Mar 2023 12:40:27 GMT\r\nContent-Type: application/json; charset=utf-8\r\nExpires: -1\r\nContent-Length: 214\r\n\r\n   at Azure.ResourceManager.Compute.ImageCollection.GetAsync(String imageName, String expand, CancellationToken cancellationToken)\r\n   at Microsoft.OneFuzz.Service.ImageOperations.GetOsInternal(Region region, String image) in /home/runner/work/onefuzz/onefuzz/src/ApiService/ApiService/onefuzzlib/ImageOperations.cs:line 140"
        ]
    },
    "image": "/subscriptions/{SUBSCRIPTION_ID}/resourceGroups/{RESOURCE_GROUP}/providers/Microsoft.Compute/galleries/{GALLERY_NAME}/images/{IMAGE_NAME}/versions/{IMAGE_VERSION}"
    ...
    "state": "creation_failed",
```

### Expected result

It should find the image provided by the Azure Compute Gallery Resource ID path.

### Actual result

The scaleset creation attempt fails and the aforementioned error message is produced.

### Other relevant information
  * I have confirmed that the app has Reader permission to the resource group, gallery, and the image.
  * I tried dropping the version information from the end of the Resource ID, and that did not solve the issue. It will result in the same error message, except rather than the message citing 'Microsoft.Compute/images/{IMAGE_VERSION}', it instead cites, 'Microsoft.Compute/images/{IMAGE_NAME}'.
  * The error message will always cite the resource group of the onefuzz instance even when the compute gallery image is in another resource group. I tried creating an compute gallery within the same resource group as onefuzz, but that did not solve the issue.

[AB#45315](https://fuzzfest.visualstudio.com/5eaacf23-2084-4d62-8fbe-7d37326a8b74/_workitems/edit/45315)

### PR DESCRIPTION
**Pull request by Porges**
- Turn `ImageReference` into its own type so it is validated early on in request submission time, and we don't end up with malformed IDs, etc.
- Add in support for shared image galleries since that was easy enough to add while I'm doing this.
- Explicitly document which image sources are permitted and how to reference them with resource IDs.

This addresses/closes #1464 for the C# port. Also fixes #2927 which was recently reported.

---

Manually verified, I can create scalesets with:

- [x] Latest image: `/subscriptions/{}/resourceGroups/{}/providers/Microsoft.Compute/galleries/{}/images/{}`
- [x] Specific image: `/subscriptions/{}/resourceGroups/{}/providers/Microsoft.Compute/galleries/{}/images/{}/versions/{}`
- [x] Marketplace image: `Canonical:0000-com-ubuntu-server-focal:20_04-lts:latest`
- [x] VM Image `/subscriptions/{}/resourceGroups/{}/providers/Microsoft.Compute/images/{}` 

Not manually verified:
- [ ] Latest shared image: `/subscriptions/{}/resourceGroups/{}/providers/Microsoft.Compute/locations/{}/sharedGalleries/{}/images/{}`
- [ ] Specific shared image: `/subscriptions/{}/resourceGroups/{}/providers/Microsoft.Compute/locations/{}/sharedGalleries/{}/images/{}/versions/{}`

### CODE REVIEW

**Comment by Porges** [comment]
Need to do some validation that shared images work as expected before merging.

**Review by tevoinea — APPROVED** [review]

**Comment by Porges** [comment]
Waiting for #2375 to merge in order to rebase this on top.

**Comment by Porges** [comment]
Doing manual verification, I can create scalesets with:

- [] `/subscriptions/{id}/resourceGroups/{rg}/providers/Microsoft.Compute/galleries/{gallery}/images/{image}`
- [] `/subscriptions/{id}/resourceGroups/{rg}/providers/Microsoft.Compute/galleries/{gallery}/images/{image}/versions/{version}`

File: src/ApiService/ApiService/onefuzzlib/ImageReference.cs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+        public sealed override Compute.Models.ImageReference ToArm()
+            => new() { Id = Identifier };
+
+        public sealed override string ToString() => Identifier.ToString();
+    }
+
+    [JsonConverter(typeof(Converter<LatestGalleryImage>))]
+    public sealed record LatestGalleryImage(ResourceIdentifier Identifier) : ArmImageReference(Identifier) {
+        public override async Task<OneFuzzResult<Os>> GetOs(ArmClient armClient, Region region) {
```
Comment (tevoinea) [inline src/ApiService/ApiService/onefuzzlib/ImageReference.cs:95]: In the previous `GetOs` implementation (in ImageOperations which is now deleted), we cached the request to Azure. We should make sure we continue caching for each of these new sub-implementations.

Comment (Porges) [reply src/ApiService/ApiService/onefuzzlib/ImageReference.cs:95]: Good call.

**Review by tevoinea — APPROVED** [review]

**Pull request merged**